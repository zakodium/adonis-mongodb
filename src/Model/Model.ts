import { Exception } from '@poppinss/utils';
import { cloneDeep, isEqual, pickBy, snakeCase } from 'lodash';
import {
  BulkWriteOptions,
  ClientSession,
  Collection,
  CountDocumentsOptions,
  CountOptions,
  DeleteOptions,
  DistinctOptions,
  Filter,
  FindOptions,
  InsertOneOptions,
} from 'mongodb';
import pluralize from 'pluralize';

import { DatabaseContract } from '@ioc:Zakodium/Mongodb/Database';
import {
  MongodbDocument,
  QueryContract,
  NoExtraProperties,
  ModelAttributes,
  ModelAdapterOptions,
  ModelDocumentOptions,
} from '@ioc:Zakodium/Mongodb/Odm';

import { proxyHandler } from './proxyHandler';

function mergeDriverOptions<
  DriverOptionType extends { session?: ClientSession | undefined },
>(options?: ModelAdapterOptions<DriverOptionType>): DriverOptionType {
  if (!options) return {} as DriverOptionType;
  return {
    ...options.driverOptions,
    session: options.client,
  } as DriverOptionType;
}

function ensureSort(options?: FindOptions): void {
  if (!options || options.sort) return;
  options.sort = {
    _id: -1,
  };
}

class Query<ModelType extends typeof BaseModel>
  implements QueryContract<InstanceType<ModelType>>
{
  public constructor(
    private filter: Filter<ModelAttributes<InstanceType<ModelType>>>,
    private options:
      | ModelAdapterOptions<
          FindOptions<ModelAttributes<InstanceType<ModelType>>>
        >
      | undefined,
    private modelConstructor: ModelType,
  ) {}

  public async first(): Promise<InstanceType<ModelType> | null> {
    const collection = await this.modelConstructor.getCollection();
    const driverOptions = mergeDriverOptions(this.options);
    ensureSort(driverOptions);
    const result = await collection.findOne(this.filter, driverOptions);
    if (result === undefined) {
      return null;
    }
    const instance = new this.modelConstructor(
      result,
      {
        // @ts-expect-error Unavoidable error.
        collection,
        session: driverOptions.session,
      },
      true,
    ) as InstanceType<ModelType>;
    return instance;
  }

  public async firstOrFail(): Promise<InstanceType<ModelType>> {
    const result = await this.first();
    if (!result) {
      throw new Exception('Document not found', 404, 'E_DOCUMENT_NOT_FOUND');
    }
    return result;
  }

  public async all(): Promise<Array<InstanceType<ModelType>>> {
    const collection = await this.modelConstructor.getCollection();
    const driverOptions = mergeDriverOptions(this.options);
    ensureSort(driverOptions);
    const result = await collection.find(this.filter, driverOptions).toArray();
    return result.map(
      (value) =>
        new this.modelConstructor(
          value,
          {
            // @ts-expect-error Unavoidable error.
            collection,
            session: driverOptions.session,
          },
          true,
        ) as InstanceType<ModelType>,
    );
  }

  public async count(): Promise<number> {
    const collection = await this.modelConstructor.getCollection();
    const driverOptions = mergeDriverOptions(this.options);
    return collection.countDocuments(
      this.filter,
      driverOptions as CountOptions,
    );
  }

  public async distinct<T = unknown>(key: string): Promise<T[]> {
    const collection = await this.modelConstructor.getCollection();
    const driverOptions = mergeDriverOptions(this.options);
    return collection.distinct(
      key,
      this.filter,
      driverOptions as DistinctOptions,
    );
  }

  public async *[Symbol.asyncIterator](): AsyncIterableIterator<
    InstanceType<ModelType>
  > {
    const collection = await this.modelConstructor.getCollection();
    const driverOptions = mergeDriverOptions(this.options);
    ensureSort(driverOptions);
    for await (const value of collection.find(this.filter, driverOptions)) {
      if (value === null) continue;
      yield new this.modelConstructor(
        value,
        {
          // @ts-expect-error Unavoidable error.
          collection,
          session: driverOptions.session,
        },
        true,
      ) as InstanceType<ModelType>;
    }
  }
}

function computeCollectionName(constructorName: string): string {
  return snakeCase(pluralize(constructorName));
}

interface InternalModelConstructorOptions {
  collection: Collection<ModelAttributes<MongodbDocument<unknown>>>;
  session?: ClientSession;
}

function hasOwn(object: unknown, key: string): boolean {
  return Object.prototype.hasOwnProperty.call(object, key);
}

interface DataToSet {
  [key: string]: unknown;
  createdAt: Date;
  updatedAt: Date;
}

export class BaseModel {
  public static connection?: string;
  public static collectionName?: string;

  public readonly _id: unknown;
  public readonly createdAt: Date;
  public readonly updatedAt: Date;

  public $isDeleted: boolean;

  protected $collection: Collection<
    ModelAttributes<MongodbDocument<unknown>>
  > | null = null;
  protected $originalData: Record<string, unknown>;
  protected $currentData: Record<string, unknown>;
  protected $options: InternalModelConstructorOptions;
  protected $alreadySaved: boolean;

  public constructor(
    dbObj?: Record<string, unknown>,
    options?: InternalModelConstructorOptions,
    alreadyExists = false,
  ) {
    if (dbObj) {
      this.$originalData = alreadyExists === true ? cloneDeep(dbObj) : {};
      this.$currentData = dbObj;
    } else {
      this.$originalData = {};
      this.$currentData = {};
    }

    if (options !== undefined) {
      this.$options = options;
      this.$collection = options.collection;
    }

    this.$alreadySaved = alreadyExists;
    this.$isDeleted = false;
    // eslint-disable-next-line no-constructor-return
    return new Proxy(this, proxyHandler);
  }

  public static $database: DatabaseContract;
  public static $setDatabase(database: DatabaseContract): void {
    this.$database = database;
  }

  public static $collectionName: string | undefined;
  public static $getCollectionName(): string {
    if (!hasOwn(this, '$collectionName')) {
      if (hasOwn(this, 'collectionName')) {
        this.$collectionName = this.collectionName;
      } else {
        this.$collectionName = computeCollectionName(this.name);
      }
    }
    return this.$collectionName as string;
  }

  public static async count<ModelType extends typeof BaseModel>(
    this: ModelType,
    filter: Filter<ModelAttributes<InstanceType<ModelType>>>,
    options?: ModelAdapterOptions<CountDocumentsOptions>,
  ): Promise<number> {
    const collection = await this.getCollection();
    const driverOptions = mergeDriverOptions(options);
    return collection.countDocuments(filter, driverOptions);
  }

  public static async create<ModelType extends typeof BaseModel>(
    this: ModelType,
    value: Partial<ModelAttributes<InstanceType<ModelType>>>,
    options?: ModelAdapterOptions<InsertOneOptions>,
  ): Promise<InstanceType<ModelType>> {
    const collection = await this.getCollection();
    const driverOptions = mergeDriverOptions(options);
    const instance = new this(value, {
      // @ts-expect-error Unavoidable error.
      collection,
      session: driverOptions.session,
    }) as InstanceType<ModelType>;
    await instance.save({ driverOptions });
    return instance;
  }

  public static async createMany<ModelType extends typeof BaseModel>(
    this: ModelType,
    values: Array<Partial<ModelAttributes<InstanceType<ModelType>>>>,
    options?: ModelAdapterOptions<BulkWriteOptions>,
  ): Promise<Array<InstanceType<ModelType>>> {
    const collection = await this.getCollection();
    const driverOptions = mergeDriverOptions(options);
    const instances = values.map(
      (value) =>
        new this(value, {
          // @ts-expect-error Unavoidable error.
          collection,
          session: driverOptions.session,
        }) as InstanceType<ModelType>,
    );
    for (const instance of instances) {
      await instance.save({ driverOptions });
    }
    return instances;
  }

  public static async find<ModelType extends typeof BaseModel>(
    this: ModelType,
    id: InstanceType<ModelType>['_id'],
    options?: ModelAdapterOptions<
      FindOptions<ModelAttributes<InstanceType<ModelType>>>
    >,
  ): Promise<InstanceType<ModelType> | null> {
    const collection = await this.getCollection();
    const driverOptions = mergeDriverOptions(options);
    const filter = { _id: id } as Filter<
      ModelAttributes<InstanceType<ModelType>>
    >;
    const result = await collection.findOne(filter, driverOptions);
    if (result === undefined) return null;
    const instance = new this(
      result,
      // @ts-expect-error Unavoidable error.
      { collection, session: driverOptions.session },
      true,
    ) as InstanceType<ModelType>;
    return instance;
  }

  public static async findOrFail<ModelType extends typeof BaseModel>(
    this: ModelType,
    id: InstanceType<ModelType>['_id'],
    options?: ModelAdapterOptions<
      FindOptions<ModelAttributes<InstanceType<ModelType>>>
    >,
  ): Promise<InstanceType<ModelType>> {
    const result = await this.find(id, options);
    if (!result) {
      throw new Exception('Document not found', 404, 'E_DOCUMENT_NOT_FOUND');
    }
    return result;
  }

  public static async findBy<ModelType extends typeof BaseModel>(
    this: ModelType,
    key: string,
    value: unknown,
    options?: ModelAdapterOptions<
      FindOptions<ModelAttributes<InstanceType<ModelType>>>
    >,
  ): Promise<InstanceType<ModelType> | null> {
    const collection = await this.getCollection();
    const driverOptions = mergeDriverOptions(options);
    const filter = { [key]: value } as Filter<
      ModelAttributes<InstanceType<ModelType>>
    >;
    const result = await collection.findOne(filter, driverOptions);
    if (result === undefined) return null;
    const instance = new this(
      result,
      // @ts-expect-error Unavoidable error.
      { collection, session: driverOptions.session },
      true,
    ) as InstanceType<ModelType>;
    return instance;
  }

  public static async findByOrFail<ModelType extends typeof BaseModel>(
    this: ModelType,
    key: string,
    value: unknown,
    options?: ModelAdapterOptions<
      FindOptions<ModelAttributes<InstanceType<ModelType>>>
    >,
  ): Promise<InstanceType<ModelType>> {
    const result = await this.findBy(key, value, options);
    if (!result) {
      throw new Exception('Document not found', 404, 'E_DOCUMENT_NOT_FOUND');
    }
    return result;
  }

  public static async findMany<ModelType extends typeof BaseModel>(
    this: ModelType,
    ids: Array<InstanceType<ModelType>['_id']>,
    options?: ModelAdapterOptions<
      FindOptions<ModelAttributes<InstanceType<ModelType>>>
    >,
  ): Promise<Array<InstanceType<ModelType>>> {
    const collection = await this.getCollection();
    const driverOptions = mergeDriverOptions(options);
    const result = await collection
      // @ts-expect-error Unavoidable error.
      .find({ _id: { $in: ids } }, driverOptions)
      .toArray();
    const instances = result.map(
      (result) =>
        new this(result, {
          // @ts-expect-error Unavoidable error.
          collection,
          session: driverOptions.session,
        }) as InstanceType<ModelType>,
    );
    return instances;
  }

  public static async all<ModelType extends typeof BaseModel>(
    this: ModelType,
    options?: ModelAdapterOptions<
      FindOptions<ModelAttributes<InstanceType<ModelType>>>
    >,
  ): Promise<Array<InstanceType<ModelType>>> {
    const collection = await this.getCollection();
    const driverOptions = mergeDriverOptions(options);
    ensureSort(driverOptions);
    const result = await collection.find({}, driverOptions).toArray();
    const instances = result.map(
      (result) =>
        new this(result, {
          // @ts-expect-error Unavoidable error.
          collection,
          session: driverOptions.session,
        }) as InstanceType<ModelType>,
    );
    return instances;
  }

  public static query<ModelType extends typeof BaseModel>(
    this: ModelType,
    filter: Filter<ModelAttributes<InstanceType<ModelType>>> = {},
    options?: ModelAdapterOptions<
      FindOptions<ModelAttributes<InstanceType<ModelType>>>
    >,
  ): Query<ModelType> {
    return new Query(filter, options, this);
  }

  public static async getCollection<ModelType extends typeof BaseModel>(
    this: ModelType,
    connection = this.connection,
  ): Promise<Collection<ModelAttributes<InstanceType<ModelType>>>> {
    if (!this.$database) {
      throw new Error('Model should only be accessed from IoC container');
    }
    const connectionInstance = this.$database.connection(connection);
    return connectionInstance.collection(this.$getCollectionName());
  }

  public [Symbol.for('nodejs.util.inspect.custom')](): unknown {
    return {
      model: this.constructor.name,
      originalData: this.$originalData,
      currentData: this.$currentData,
      isDirty: this.$isDirty,
    };
  }

  public get $dirty(): Partial<ModelAttributes<this>> {
    return pickBy(this.$currentData, (value, key) => {
      return (
        this.$originalData[key] === undefined ||
        !isEqual(this.$originalData[key], value)
      );
    }) as Partial<ModelAttributes<this>>;
  }

  public $ensureNotDeleted(): void {
    if (this.$isDeleted) {
      throw new Exception('Document was deleted', 500, 'E_DOCUMENT_DELETED');
    }
  }

  public async $ensureCollection(): Promise<
    Collection<ModelAttributes<MongodbDocument<unknown>>>
  > {
    if (this.$collection !== null) {
      return this.$collection;
    }

    const constructor = this.constructor as typeof BaseModel;
    this.$collection =
      (await constructor.getCollection()) as unknown as Collection<
        ModelAttributes<MongodbDocument<unknown>>
      >;
    return this.$collection;
  }

  public $prepareToSet() {
    const dirty = this.$dirty;
    const dirtyEntries = Object.entries(dirty);
    if (dirtyEntries.length === 0) {
      return null;
    }

    // We cheat a little bit with the assertion. This is necessary because the
    // value returned by this function can be used in a MongoDB update query
    // which shouldn't reset the createdAt field.
    const toSet = {} as DataToSet;
    const now = new Date();
    if (this.$currentData.createdAt === undefined) {
      this.$currentData.createdAt = now;
      toSet.createdAt = now;
    }
    this.$currentData.updatedAt = now;
    toSet.updatedAt = now;

    for (const [dirtyKey, dirtyValue] of dirtyEntries) {
      toSet[dirtyKey] = dirtyValue;
    }
    return toSet;
  }

  public get id() {
    return this.$currentData._id;
  }

  public get $isDirty(): boolean {
    return Object.keys(this.$dirty).length > 0;
  }

  public toJSON(): unknown {
    return this.$currentData;
  }

  public async save(
    options?: ModelDocumentOptions<InsertOneOptions>,
  ): Promise<boolean> {
    this.$ensureNotDeleted();
    const collection = await this.$ensureCollection();

    const toSet = this.$prepareToSet();
    if (toSet === null) return false;
    const driverOptions = {
      ...options?.driverOptions,
      session: this.$options?.session,
    };
    if (this.$alreadySaved === false) {
      const result = await collection.insertOne(toSet, driverOptions);
      this.$currentData._id = result.insertedId;
    } else {
      await collection.updateOne(
        { _id: this.$currentData._id },
        { $set: toSet },
        driverOptions,
      );
    }
    this.$originalData = cloneDeep(this.$currentData);
    this.$alreadySaved = true;
    return true;
  }

  public async delete(
    options?: ModelDocumentOptions<DeleteOptions>,
  ): Promise<boolean> {
    this.$ensureNotDeleted();
    const collection = await this.$ensureCollection();
    const driverOptions = {
      ...options?.driverOptions,
      session: this.$options?.session,
    };
    const result = await collection.deleteOne(
      {
        _id: this.$currentData._id,
      },
      driverOptions,
    );
    this.$isDeleted = true;
    return result.deletedCount === 1;
  }

  public merge<T extends Partial<ModelAttributes<this>>>(
    values: NoExtraProperties<Partial<ModelAttributes<this>>, T>,
  ): this {
    Object.entries(values).forEach(([key, value]) => {
      this.$currentData[key] = value;
    });
    return this;
  }

  public fill<T extends Partial<ModelAttributes<this>>>(
    values: NoExtraProperties<Partial<ModelAttributes<this>>, T>,
  ) {
    const createdAt = this.$currentData.createdAt;
    this.$currentData = {
      _id: this.id,
    };
    if (createdAt) this.$currentData.createdAt = createdAt;
    return this.merge(values);
  }
}

export class BaseAutoIncrementModel extends BaseModel {
  public readonly _id: number;

  public async save(
    options?: ModelDocumentOptions<InsertOneOptions>,
  ): Promise<boolean> {
    this.$ensureNotDeleted();
    const collection = await this.$ensureCollection();

    const toSet = this.$prepareToSet();
    if (toSet === null) return false;

    const driverOptions = {
      ...options?.driverOptions,
      session: this.$options?.session,
    };

    if (this._id === undefined) {
      const connection = BaseAutoIncrementModel.$database.connection();
      const counterCollection = await connection.collection<{ count: number }>(
        '__adonis_mongodb_counters',
      );

      const doc = await counterCollection.findOneAndUpdate(
        { _id: computeCollectionName(this.constructor.name) },
        { $inc: { count: 1 } },
        { ...driverOptions, upsert: true },
      );
      const newCount = doc.value ? doc.value.count + 1 : 1;
      toSet._id = newCount;
      await collection.insertOne(toSet, driverOptions);
      this.$currentData._id = newCount;
    } else {
      await collection.updateOne(
        { _id: this.$currentData._id },
        { $set: toSet },
        driverOptions,
      );
    }
    this.$originalData = cloneDeep(this.$currentData);
    return true;
  }
}
