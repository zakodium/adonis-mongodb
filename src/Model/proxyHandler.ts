/* eslint-disable @typescript-eslint/no-explicit-any */
import { BaseModel } from './Model';

export const proxyHandler: ProxyHandler<any> = {
  get(target: any, prop: string | symbol, receiver: any) {
    const Model = target.constructor as typeof BaseModel;
    if (Model.$hasComputed(prop as string)) {
      const property = Object.getOwnPropertyDescriptor(
        Object.getPrototypeOf(target),
        prop,
      );
      if (property?.get) {
        return property.get.call(target.$attributes);
      }
    }

    if (target[prop] !== undefined) {
      return Reflect.get(target, prop, receiver);
    }

    return Reflect.get(target.$attributes, prop, receiver);
  },
  set(target: any, prop: string | symbol, value: any) {
    if (target[prop] !== undefined) {
      return Reflect.set(target, prop, value);
    }
    return Reflect.set(target.$attributes, prop, value);
  },
  ownKeys() {
    throw new Error(
      'Getting model keys is disallowed. If you want to use object spread on the current data, do { ...model.$attributes }',
    );
  },
};
