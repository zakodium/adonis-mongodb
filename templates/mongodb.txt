import Env from '@ioc:Adonis/Core/Env';
import { MongodbConfig } from '@ioc:Zakodium/Mongodb/Database';

const mongodbConfig: MongodbConfig = {
  connection: Env.get('MONGODB_CONNECTION'),
  connections: {
    mongodb: {
      url: Env.get('MONGODB_URL'),
      database: Env.get('MONGODB_DATABASE'),
    },
  },
};

export default mongodbConfig;
