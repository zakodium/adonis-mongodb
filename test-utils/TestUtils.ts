import path from 'node:path';

import { FakeLogger } from '@adonisjs/logger';

import type { MongodbConfig } from '@ioc:Zakodium/Mongodb/Database';

import { Connection } from '../src/Database/Connection';
import { Database } from '../src/Database/Database';
import { BaseAutoIncrementModel, BaseModel } from '../src/Model/Model';

export function getLogger() {
  const loggerConfig = {
    name: 'adonis-logger',
    level: 'trace',
    messageKey: 'msg',
    enabled: true,
  };
  return new FakeLogger(loggerConfig);
}

export function getConnection(logger = getLogger()) {
  const connectionConfig = {
    url: 'mongodb://localhost:33333?directConnection=true',
    database: `test-runner-${path.basename(
      expect.getState().testPath as string,
      '.test.ts',
    )}`,
  };
  return new Connection('mongo', connectionConfig, logger);
}

export function getMongodb(logger = getLogger()) {
  const database = `test-runner-${path
    .basename(expect.getState().testPath as string, '.test.ts')
    .replaceAll('.', '_')}`;
  const mongoConfig: MongodbConfig = {
    connection: 'mongo',
    connections: {
      mongo: {
        url: 'mongodb://127.0.0.1:33333?directConnection=true',
        database,
      },
      other: {
        url: 'mongodb://127.0.0.1:33333?directConnection=true',
        database,
      },
    },
  };

  return new Database(mongoConfig, logger);
}

export function setupDatabase() {
  const db = getMongodb();
  BaseModel.$setDatabase(db);
  BaseAutoIncrementModel.$setDatabase(db);

  afterAll(async () => {
    const database = await db.connection('mongo').database();
    await database.dropDatabase();
    await db.manager.closeAll();
  });

  return db;
}
