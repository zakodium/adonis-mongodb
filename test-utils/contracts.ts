declare module '@ioc:Adonis/Core/Hash' {
  interface HashersList {
    bcrypt: HashDrivers['bcrypt'];
  }
}
