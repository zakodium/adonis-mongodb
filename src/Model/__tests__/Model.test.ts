import { ObjectId } from 'mongodb';

import { getMongodb } from '../../../test-utils/TestUtils';
import { Model, AutoIncrementModel } from '../Model';

interface IUser {
  username: string;
  password: string;
}

interface IPost {
  title: string;
  content: string;
}

class User extends Model implements IUser {
  public static collectionName = 'users';

  public username: string;
  public password: string;
}

class Post extends AutoIncrementModel implements IPost {
  public static collectionName = 'posts';

  public title: string;
  public content: string;
}

class Something extends Model {
  public static collectionName = 'somethingElse';

  public test: boolean;
}

let usernameCounter = 0;
function nextUsername() {
  return `root${++usernameCounter}`;
}

let postTitleCounter = 0;
function nextTitle() {
  return `post title ${++postTitleCounter}`;
}

const db = getMongodb();
Model.$setDatabase(db);

/*
afterAll(async () => {
  await (await db.connection('mongo').database()).dropDatabase();
  await db.closeConnections();
});*/

beforeAll(async () => {
  await (await db.connection('mongo').database()).dropDatabase();
});

test('can create', async () => {
  const newUser = await User.create({
    username: nextUsername(),
    password: 'root',
  });
  await newUser.save();
  expect(newUser).toBeDefined();
});

test('get collection by model class', async () => {
  const collection = await User.getCollection();
  expect(collection).toBeDefined();
});

test('find one by property', async () => {
  const user = User.findOne({ username: 'root', password: 'root' });
  expect(user).toBeDefined();
});

test('find all', async () => {
  await User.create({
    username: nextUsername(),
    password: 'root',
  });

  const users = await User.find({});
  expect(await users.count()).toBe(2);
});

test('find by id should work', async () => {
  const user = await User.create({
    username: nextUsername(),
    password: 'root',
  });
  const secondUser = await User.findById(user.id);
  expect(secondUser).not.toBeNull();
});

test("find by id should throw when doesn't exists", async () => {
  const t = async () => {
    await User.findByIdOrThrow('notavalidid');
  };
  await expect(t).rejects.toStrictEqual(
    new Error('document notavalidid not found in users'),
  );
});

test('saved changes should be sent to database', async () => {
  const user = await User.create({
    username: nextUsername(),
    password: 'root',
  });
  user.password = 'rootroot';
  await user.save();

  const sameUser = await User.findById(user.id);
  expect(sameUser).not.toBeNull();
  expect((sameUser as User).password).toStrictEqual('rootroot');
});

test('id is an ObjectId', async () => {
  const user = await User.create({
    username: nextUsername(),
    password: 'root',
  });
  await user.save();

  expect(user.id).toBeInstanceOf(ObjectId);
});

test('delete on model', async () => {
  const user = await User.create({
    username: nextUsername(),
    password: 'root',
  });

  await user.delete();

  const sameUserButDeleted = await User.findById(user.id);
  expect(sameUserButDeleted).toBeNull();
});

test('id is a number on AutoIncrementModel', async () => {
  const firstPost = await Post.create({
    title: nextTitle(),
    content: 'post content',
  });
  expect(firstPost.id).toBe(1);
  expect(typeof firstPost.id).toBe('number');
});

test('AutoIncrementModel id increments', async () => {
  const firstPost = await Post.create({
    title: nextTitle(),
    content: 'post content',
  });
  const secondPost = await Post.create({
    title: nextTitle(),
    content: 'post content',
  });
  expect(firstPost.id).toBe(secondPost.id - 1);
});

test('passing session should run requests within the same session', async () => {
  const username = nextUsername();
  await db.connection('mongo').transaction(async (session) => {
    const user = await User.create(
      {
        username: username,
        password: 'rootroot',
      },
      { session },
    );

    user.password = 'root';

    await user.save();

    const shouldNotExist = await User.findOne({ username: username });
    expect(shouldNotExist).toBeNull();
  });

  const shouldExistNow = await User.findOne({ username: 'root7' });
  expect(shouldExistNow).not.toBeNull();
  expect(shouldExistNow?.password).toBe('root');
});

test('class instantiation Model should create an entry', async () => {
  const user = new User();
  user.username = nextUsername();
  user.password = 'rootroot';
  await user.save();

  const shouldExist = await User.findOne({ username: 'root8' });
  expect(shouldExist).not.toBeNull();
  expect(user.id).toBeInstanceOf(ObjectId);
});

test('class instantiation Model should be updatable', async () => {
  const username = nextUsername();
  const user = new User();
  user.username = username;
  user.password = 'rootroot';
  await user.save();

  user.password = 'root';
  await user.save();

  const shouldHaveNewPassword = await User.findOne({ username });
  expect(shouldHaveNewPassword?.password).toBe('root');
});

test('find one returns should not be dirty', async () => {
  const username = nextUsername();
  await User.create({
    username,
    password: 'rootroot',
  });

  const foundUser = await User.findOne({ username });
  expect(foundUser?.isDirty).toBe(false);
});

test('class instantiation auto incremented model', async () => {
  const post = new Post();
  post.title = nextTitle();
  post.content = 'post content';
  await post.save();

  expect(typeof post.id).toBe('number');
});

test('custom collection name - class', async () => {
  const something = await Something.create({ test: false });
  await something.save();
  expect((await Something.getCollection()).collectionName).toBe(
    Something.collectionName,
  );
});

test('custom collection name - instance', async () => {
  const something = new Something();
  something.test = true;
  await something.save();

  const found = await (
    await Model.$database.connection().collection(Something.collectionName)
  ).findOne({ _id: something.id });

  expect(found).not.toBeNull();
});
