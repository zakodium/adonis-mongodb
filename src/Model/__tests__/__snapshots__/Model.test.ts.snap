// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom inspect function 1`] = `
"{
  Model: 'Post',
  '$original': { _id: 4, title: 'post title 4', content: 'post content' },
  '$attributes': { _id: 4, title: 'post title 4', content: 'new content' },
  '$isPersisted': true,
  '$isNew': false,
  '$isLocal': false,
  '$isDeleted': false,
  '$dirty': { content: 'new content' },
  '$isDirty': true
}"
`;
